File: docs/examples/python/convert.py
#!/usr/bin/env python3
"""
XToPDF API - Python Usage Examples

Requirements:
    pip install requests

Usage:
    python convert.py
"""

import requests
import time
import sys

BASE_URL = "http://localhost:8080"


def basic_convert(file_path: str, output_name: str = "output.pdf") -> bytes:
    """Convert a single file to PDF."""
    with open(file_path, "rb") as f:
        response = requests.post(
            f"{BASE_URL}/api/convert",
            files={"file": (file_path.split("/")[-1], f)},
            params={"outputFileName": output_name},
        )

    if response.status_code == 200:
        print(f"Conversion successful: {len(response.content)} bytes")
        return response.content
    else:
        print(f"Conversion failed: {response.status_code}")
        print(response.json())
        sys.exit(1)


def batch_convert(file_paths: list[str]) -> bytes:
    """Convert multiple files to PDF in a single batch request."""
    files = []
    for path in file_paths:
        with open(path, "rb") as f:
            files.append(("files", (path.split("/")[-1], f.read())))

    response = requests.post(
        f"{BASE_URL}/api/convert/batch",
        files=files,
    )

    if response.status_code == 200:
        print(f"Batch conversion successful: {len(response.content)} bytes")
        return response.content
    else:
        print(f"Batch conversion failed: {response.status_code}")
        print(response.json())
        sys.exit(1)


def async_convert(file_path: str, webhook_url: str = None) -> dict:
    """Submit an async conversion job and poll for completion."""
    with open(file_path, "rb") as f:
        params = {}
        if webhook_url:
            params["webhookUrl"] = webhook_url

        response = requests.post(
            f"{BASE_URL}/api/convert/async",
            files={"file": (file_path.split("/")[-1], f)},
            params=params,
        )

    if response.status_code != 202:
        print(f"Job submission failed: {response.status_code}")
        print(response.json())
        sys.exit(1)

    job = response.json()
    job_id = job["jobId"]
    print(f"Job submitted: {job_id}")

    # Poll for completion
    while True:
        status_response = requests.get(f"{BASE_URL}/api/convert/async/{job_id}")
        status = status_response.json()

        print(f"  Status: {status['status']}")

        if status["status"] == "COMPLETED":
            # Download the result
            result_response = requests.get(
                f"{BASE_URL}/api/convert/async/{job_id}/result"
            )
            print(f"Result downloaded: {len(result_response.content)} bytes")
            return {"job": status, "content": result_response.content}

        elif status["status"] == "FAILED":
            print(f"Job failed: {status.get('errorMessage', 'Unknown error')}")
            sys.exit(1)

        time.sleep(2)


def add_watermark(pdf_path: str, text: str, layer: str = "FOREGROUND") -> bytes:
    """Add a watermark to an existing PDF."""
    with open(pdf_path, "rb") as f:
        response = requests.post(
            f"{BASE_URL}/api/pdf/watermark",
            files={"file": (pdf_path.split("/")[-1], f)},
            params={"text": text, "layer": layer},
        )

    if response.status_code == 200:
        print(f"Watermark added: {len(response.content)} bytes")
        return response.content
    else:
        print(f"Watermark failed: {response.status_code}")
        print(response.json())
        sys.exit(1)


def add_page_numbers(
    pdf_path: str,
    position: str = "BOTTOM",
    alignment: str = "CENTER",
    style: str = "ARABIC",
) -> bytes:
    """Add page numbers to an existing PDF."""
    with open(pdf_path, "rb") as f:
        response = requests.post(
            f"{BASE_URL}/api/pdf/page-numbers",
            files={"file": (pdf_path.split("/")[-1], f)},
            params={
                "position": position,
                "alignment": alignment,
                "style": style,
            },
        )

    if response.status_code == 200:
        print(f"Page numbers added: {len(response.content)} bytes")
        return response.content
    else:
        print(f"Page numbers failed: {response.status_code}")
        print(response.json())
        sys.exit(1)


def merge_pdfs(pdf_path: str, overlay_path: str, position: str = "back") -> bytes:
    """Merge two PDF files."""
    with open(pdf_path, "rb") as f1, open(overlay_path, "rb") as f2:
        response = requests.post(
            f"{BASE_URL}/api/pdf/merge",
            files={
                "file": (pdf_path.split("/")[-1], f1),
                "overlayFile": (overlay_path.split("/")[-1], f2),
            },
            params={"position": position},
        )

    if response.status_code == 200:
        print(f"Merge successful: {len(response.content)} bytes")
        return response.content
    else:
        print(f"Merge failed: {response.status_code}")
        print(response.json())
        sys.exit(1)


if __name__ == "__main__":
    print("=== XToPDF Python Examples ===\n")

    # Example: Basic conversion
    print("1. Basic conversion:")
    print('   pdf = basic_convert("document.docx")')
    print()

    # Example: Batch conversion
    print("2. Batch conversion:")
    print('   pdf = batch_convert(["file1.docx", "file2.xlsx", "file3.png"])')
    print()

    # Example: Async conversion
    print("3. Async conversion with polling:")
    print('   result = async_convert("large-file.xlsx")')
    print()

    # Example: Async with webhook
    print("4. Async conversion with webhook:")
    print('   result = async_convert("large-file.xlsx", webhook_url="https://myapp.com/webhook")')
    print()

    # Example: PDF operations
    print("5. Add watermark:")
    print('   pdf = add_watermark("document.pdf", "CONFIDENTIAL")')
    print()

    print("6. Add page numbers:")
    print('   pdf = add_page_numbers("document.pdf", position="BOTTOM", alignment="CENTER")')
    print()

    print("7. Merge PDFs:")
    print('   pdf = merge_pdfs("main.pdf", "appendix.pdf", position="back")')

File: docs/examples/javascript/convert.js
/**
 * XToPDF API - JavaScript (Node.js) Usage Examples
 *
 * Requirements:
 *   Node.js 18+ (native fetch API)
 *
 * Usage:
 *   node convert.js
 */

const fs = require("fs");
const path = require("path");

const BASE_URL = "http://localhost:8080";

/**
 * Convert a single file to PDF.
 */
async function basicConvert(filePath, outputName = "output.pdf") {
  const file = fs.readFileSync(filePath);
  const formData = new FormData();
  formData.append("file", new Blob([file]), path.basename(filePath));

  const response = await fetch(
    `${BASE_URL}/api/convert?outputFileName=${encodeURIComponent(outputName)}`,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`Conversion failed (${response.status}): ${error.message}`);
  }

  const pdfBuffer = Buffer.from(await response.arrayBuffer());
  console.log(`Conversion successful: ${pdfBuffer.length} bytes`);
  return pdfBuffer;
}

/**
 * Convert multiple files in a single batch request.
 */
async function batchConvert(filePaths) {
  const formData = new FormData();

  for (const filePath of filePaths) {
    const file = fs.readFileSync(filePath);
    formData.append("files", new Blob([file]), path.basename(filePath));
  }

  const response = await fetch(`${BASE_URL}/api/convert/batch`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(
      `Batch conversion failed (${response.status}): ${error.message}`
    );
  }

  const resultBuffer = Buffer.from(await response.arrayBuffer());
  console.log(`Batch conversion successful: ${resultBuffer.length} bytes`);
  return resultBuffer;
}

/**
 * Submit an async conversion job and poll for completion.
 */
async function asyncConvert(filePath, webhookUrl = null) {
  const file = fs.readFileSync(filePath);
  const formData = new FormData();
  formData.append("file", new Blob([file]), path.basename(filePath));

  let url = `${BASE_URL}/api/convert/async`;
  if (webhookUrl) {
    url += `?webhookUrl=${encodeURIComponent(webhookUrl)}`;
  }

  const submitResponse = await fetch(url, {
    method: "POST",
    body: formData,
  });

  if (submitResponse.status !== 202) {
    const error = await submitResponse.json();
    throw new Error(
      `Job submission failed (${submitResponse.status}): ${error.message}`
    );
  }

  const job = await submitResponse.json();
  const jobId = job.jobId;
  console.log(`Job submitted: ${jobId}`);

  // Poll for completion
  while (true) {
    const statusResponse = await fetch(
      `${BASE_URL}/api/convert/async/${jobId}`
    );
    const status = await statusResponse.json();
    console.log(`  Status: ${status.status}`);

    if (status.status === "COMPLETED") {
      const resultResponse = await fetch(
        `${BASE_URL}/api/convert/async/${jobId}/result`
      );
      const pdfBuffer = Buffer.from(await resultResponse.arrayBuffer());
      console.log(`Result downloaded: ${pdfBuffer.length} bytes`);
      return { job: status, content: pdfBuffer };
    }

    if (status.status === "FAILED") {
      throw new Error(`Job failed: ${status.errorMessage || "Unknown error"}`);
    }

    await new Promise((resolve) => setTimeout(resolve, 2000));
  }
}

/**
 * Add a watermark to an existing PDF.
 */
async function addWatermark(pdfPath, text, layer = "FOREGROUND") {
  const file = fs.readFileSync(pdfPath);
  const formData = new FormData();
  formData.append("file", new Blob([file]), path.basename(pdfPath));

  const params = new URLSearchParams({ text, layer });
  const response = await fetch(
    `${BASE_URL}/api/pdf/watermark?${params.toString()}`,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`Watermark failed (${response.status}): ${error.message}`);
  }

  const pdfBuffer = Buffer.from(await response.arrayBuffer());
  console.log(`Watermark added: ${pdfBuffer.length} bytes`);
  return pdfBuffer;
}

/**
 * Add page numbers to an existing PDF.
 */
async function addPageNumbers(
  pdfPath,
  position = "BOTTOM",
  alignment = "CENTER",
  style = "ARABIC"
) {
  const file = fs.readFileSync(pdfPath);
  const formData = new FormData();
  formData.append("file", new Blob([file]), path.basename(pdfPath));

  const params = new URLSearchParams({ position, alignment, style });
  const response = await fetch(
    `${BASE_URL}/api/pdf/page-numbers?${params.toString()}`,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(
      `Page numbers failed (${response.status}): ${error.message}`
    );
  }

  const pdfBuffer = Buffer.from(await response.arrayBuffer());
  console.log(`Page numbers added: ${pdfBuffer.length} bytes`);
  return pdfBuffer;
}

/**
 * Merge two PDF files.
 */
async function mergePdfs(pdfPath, overlayPath, position = "back") {
  const file1 = fs.readFileSync(pdfPath);
  const file2 = fs.readFileSync(overlayPath);
  const formData = new FormData();
  formData.append("file", new Blob([file1]), path.basename(pdfPath));
  formData.append("overlayFile", new Blob([file2]), path.basename(overlayPath));

  const response = await fetch(
    `${BASE_URL}/api/pdf/merge?position=${position}`,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`Merge failed (${response.status}): ${error.message}`);
  }

  const pdfBuffer = Buffer.from(await response.arrayBuffer());
  console.log(`Merge successful: ${pdfBuffer.length} bytes`);
  return pdfBuffer;
}

// --- Main ---
async function main() {
  console.log("=== XToPDF JavaScript Examples ===\n");

  console.log("1. Basic conversion:");
  console.log('   const pdf = await basicConvert("document.docx");');
  console.log('   fs.writeFileSync("output.pdf", pdf);\n');

  console.log("2. Batch conversion:");
  console.log(
    '   const zip = await batchConvert(["file1.docx", "file2.xlsx"]);\n'
  );

  console.log("3. Async conversion:");
  console.log('   const result = await asyncConvert("large-file.xlsx");\n');

  console.log("4. Async with webhook:");
  console.log(
    '   const result = await asyncConvert("large.xlsx", "https://myapp.com/hook");\n'
  );

  console.log("5. Add watermark:");
  console.log(
    '   const pdf = await addWatermark("doc.pdf", "CONFIDENTIAL");\n'
  );

  console.log("6. Add page numbers:");
  console.log("   const pdf = await addPageNumbers(\"doc.pdf\");\n");

  console.log("7. Merge PDFs:");
  console.log('   const pdf = await mergePdfs("main.pdf", "appendix.pdf");\n');
}

main().catch(console.error);

module.exports = {
  basicConvert,
  batchConvert,
  asyncConvert,
  addWatermark,
  addPageNumbers,
  mergePdfs,
};
